### ISSUE TITLE
Treat primitives and their autoboxed types as one and the same

### ISSUE THREAD

**Issue by llorllale** [yegor256/jpeek#100]
Following on from [this](https://github.com/yegor256/jpeek/issues/53#issuecomment-341963890) comment:

> I think we should treat Integer and int as one type.

Each instance of a primitive and its respective autoboxed type should count as the same type. For example:

```java
public String doSomething(int id, Integer other);
```
We should coalesce `int` and `Integer` into a single type for analyses, so our metrics should calculate the total parameter types (both input and output) for this method as **2** instead of **3**.

**Comment by 0crat** [comment]
Job [#100](https://github.com/yegor256/jpeek/issues/100) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
Bug was reported, see [§29](http://datum.zerocracy.com/pages/policy.html#29): +15 points just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale), total is [+165](https://www.0crat.com/u/llorllale)

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
java.lang.IllegalStateException: BkParallel-9 interrupted while waiting for "claims.xml" in C7JGJ00DP for 55s
	at com.zerocracy.farm.sync.SyncProject.acq(SyncProject.java:88)
	at com.zerocracy.farm.strict.StrictProject.acq(StrictProject.java:121)
	at com.zerocracy.farm.props.PropsProject.acq(PropsProject.java:78)
	at com.zerocracy.farm.footprint.FtProject.acq(FtProject.java:67)
	at com.zerocracy.farm.uplinked.UplinkedProject.lambda$acq$4(UplinkedProject.java:99)
	at org.cactoos.scalar.Ternary.value(Ternary.java:123)
	at org.cactoos.func.IoCheckedFunc.lambda$apply$0(IoCheckedFunc.java:60)
	at org.cactoos.func.IoCheckedBiFunc.apply(IoCheckedBiFunc.java:66)
	at org.cactoos.func.IoCheckedFunc.apply(IoCheckedFunc.java:61)
	at org.cactoos.scalar.IoCheckedScalar.value(IoCheckedScalar.java:65)
	at com.zerocracy.farm.uplinked.UplinkedProject.acq(UplinkedProject.java:101)
	at com.zerocracy.farm.ruled.RdProject.acq(RdProject.java:54)
	at com.zerocracy.farm.reactive.RvProject.acq(RvProject.java...
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm/tree/1.0-SNAPSHOT) /cc @yegor256

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
java.lang.IllegalStateException: BkParallel-3 interrupted while waiting for "claims.xml" in C7JGJ00DP for 55s
	at com.zerocracy.farm.sync.SyncProject.acq(SyncProject.java:88)
	at com.zerocracy.farm.strict.StrictProject.acq(StrictProject.java:121)
	at com.zerocracy.farm.props.PropsProject.acq(PropsProject.java:78)
	at com.zerocracy.farm.footprint.FtProject.acq(FtProject.java:67)
	at com.zerocracy.farm.uplinked.UplinkedProject.lambda$acq$4(UplinkedProject.java:99)
	at org.cactoos.scalar.Ternary.value(Ternary.java:123)
	at org.cactoos.func.IoCheckedFunc.lambda$apply$0(IoCheckedFunc.java:60)
	at org.cactoos.func.IoCheckedBiFunc.apply(IoCheckedBiFunc.java:66)
	at org.cactoos.func.IoCheckedFunc.apply(IoCheckedFunc.java:61)
	at org.cactoos.scalar.IoCheckedScalar.value(IoCheckedScalar.java:65)
	at com.zerocracy.farm.uplinked.UplinkedProject.acq(UplinkedProject.java:101)
	at com.zerocracy.farm.ruled.RdProject.acq(RdProject.java:54)
	at com.zerocracy.farm.reactive.RvProject.acq(RvProject.java...
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm/tree/1.0-SNAPSHOT) /cc @yegor256

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
java.lang.IllegalStateException: BkParallel-3 interrupted while waiting for "claims.xml" in C7JGJ00DP for 56s
	at com.zerocracy.farm.sync.SyncProject.acq(SyncProject.java:88)
	at com.zerocracy.farm.strict.StrictProject.acq(StrictProject.java:121)
	at com.zerocracy.farm.props.PropsProject.acq(PropsProject.java:78)
	at com.zerocracy.farm.footprint.FtProject.acq(FtProject.java:67)
	at com.zerocracy.farm.uplinked.UplinkedProject.lambda$acq$4(UplinkedProject.java:99)
	at org.cactoos.scalar.Ternary.value(Ternary.java:123)
	at org.cactoos.func.IoCheckedFunc.lambda$apply$0(IoCheckedFunc.java:60)
	at org.cactoos.func.IoCheckedBiFunc.apply(IoCheckedBiFunc.java:66)
	at org.cactoos.func.IoCheckedFunc.apply(IoCheckedFunc.java:61)
	at org.cactoos.scalar.IoCheckedScalar.value(IoCheckedScalar.java:65)
	at com.zerocracy.farm.uplinked.UplinkedProject.acq(UplinkedProject.java:101)
	at com.zerocracy.farm.ruled.RdProject.acq(RdProject.java:54)
	at com.zerocracy.farm.reactive.RvProject.acq(RvProject.java...
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm/tree/1.0-SNAPSHOT) /cc @yegor256

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
java.lang.IllegalStateException: BkParallel-1 interrupted while waiting for "claims.xml" in C7JGJ00DP for 55s
	at com.zerocracy.farm.sync.SyncProject.acq(SyncProject.java:88)
	at com.zerocracy.farm.strict.StrictProject.acq(StrictProject.java:121)
	at com.zerocracy.farm.props.PropsProject.acq(PropsProject.java:78)
	at com.zerocracy.farm.footprint.FtProject.acq(FtProject.java:67)
	at com.zerocracy.farm.uplinked.UplinkedProject.lambda$acq$4(UplinkedProject.java:99)
	at org.cactoos.scalar.Ternary.value(Ternary.java:123)
	at org.cactoos.func.IoCheckedFunc.lambda$apply$0(IoCheckedFunc.java:60)
	at org.cactoos.func.IoCheckedBiFunc.apply(IoCheckedBiFunc.java:66)
	at org.cactoos.func.IoCheckedFunc.apply(IoCheckedFunc.java:61)
	at org.cactoos.scalar.IoCheckedScalar.value(IoCheckedScalar.java:65)
	at com.zerocracy.farm.uplinked.UplinkedProject.acq(UplinkedProject.java:101)
	at com.zerocracy.farm.ruled.RdProject.acq(RdProject.java:54)
	at com.zerocracy.farm.reactive.RvProject.acq(RvProject.java...
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm/tree/1.0-SNAPSHOT) /cc @yegor256

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
java.lang.IllegalStateException: BkParallel-7 interrupted while waiting for "claims.xml" in C7JGJ00DP for 55s
	at com.zerocracy.farm.sync.SyncProject.acq(SyncProject.java:88)
	at com.zerocracy.farm.strict.StrictProject.acq(StrictProject.java:121)
	at com.zerocracy.farm.props.PropsProject.acq(PropsProject.java:78)
	at com.zerocracy.farm.footprint.FtProject.acq(FtProject.java:67)
	at com.zerocracy.farm.uplinked.UplinkedProject.lambda$acq$4(UplinkedProject.java:99)
	at org.cactoos.scalar.Ternary.value(Ternary.java:123)
	at org.cactoos.func.IoCheckedFunc.lambda$apply$0(IoCheckedFunc.java:60)
	at org.cactoos.func.IoCheckedBiFunc.apply(IoCheckedBiFunc.java:66)
	at org.cactoos.func.IoCheckedFunc.apply(IoCheckedFunc.java:61)
	at org.cactoos.scalar.IoCheckedScalar.value(IoCheckedScalar.java:65)
	at com.zerocracy.farm.uplinked.UplinkedProject.acq(UplinkedProject.java:101)
	at com.zerocracy.farm.ruled.RdProject.acq(RdProject.java:54)
	at com.zerocracy.farm.reactive.RvProject.acq(RvProject.java...
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm/tree/1.0-SNAPSHOT) /cc @yegor256

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
java.lang.IllegalStateException: BkParallel-9 interrupted while waiting for "claims.xml" in C7JGJ00DP for 54s
	at com.zerocracy.farm.sync.SyncProject.acq(SyncProject.java:88)
	at com.zerocracy.farm.strict.StrictProject.acq(StrictProject.java:121)
	at com.zerocracy.farm.props.PropsProject.acq(PropsProject.java:78)
	at com.zerocracy.farm.footprint.FtProject.acq(FtProject.java:67)
	at com.zerocracy.farm.uplinked.UplinkedProject.lambda$acq$4(UplinkedProject.java:99)
	at org.cactoos.scalar.Ternary.value(Ternary.java:123)
	at org.cactoos.func.IoCheckedFunc.lambda$apply$0(IoCheckedFunc.java:60)
	at org.cactoos.func.IoCheckedBiFunc.apply(IoCheckedBiFunc.java:66)
	at org.cactoos.func.IoCheckedFunc.apply(IoCheckedFunc.java:61)
	at org.cactoos.scalar.IoCheckedScalar.value(IoCheckedScalar.java:65)
	at com.zerocracy.farm.uplinked.UplinkedProject.acq(UplinkedProject.java:101)
	at com.zerocracy.farm.ruled.RdProject.acq(RdProject.java:54)
	at com.zerocracy.farm.reactive.RvProject.acq(RvProject.java...
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm/tree/1.0-SNAPSHOT) /cc @yegor256

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
java.lang.IllegalStateException: BkParallel-4 interrupted while waiting for "claims.xml" in C7JGJ00DP for 55s
	at com.zerocracy.farm.sync.SyncProject.acq(SyncProject.java:88)
	at com.zerocracy.farm.strict.StrictProject.acq(StrictProject.java:121)
	at com.zerocracy.farm.props.PropsProject.acq(PropsProject.java:78)
	at com.zerocracy.farm.footprint.FtProject.acq(FtProject.java:67)
	at com.zerocracy.farm.uplinked.UplinkedProject.lambda$acq$4(UplinkedProject.java:99)
	at org.cactoos.scalar.Ternary.value(Ternary.java:123)
	at org.cactoos.func.IoCheckedFunc.lambda$apply$0(IoCheckedFunc.java:60)
	at org.cactoos.func.IoCheckedBiFunc.apply(IoCheckedBiFunc.java:66)
	at org.cactoos.func.IoCheckedFunc.apply(IoCheckedFunc.java:61)
	at org.cactoos.scalar.IoCheckedScalar.value(IoCheckedScalar.java:65)
	at com.zerocracy.farm.uplinked.UplinkedProject.acq(UplinkedProject.java:101)
	at com.zerocracy.farm.ruled.RdProject.acq(RdProject.java:54)
	at com.zerocracy.farm.reactive.RvProject.acq(RvProject.java...
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm/tree/1.0-SNAPSHOT) /cc @yegor256

**Comment by amihaiemil** [comment]
@0crat status

**Comment by 0crat** [comment]
Are you speaking to me or about me [here](https://github.com/yegor256/jpeek/issues/100#issuecomment-362893584). You must always start your message with my name if you want to address it to me, see [§1](http://datum.zerocracy.com/pages/policy.html#1).

**Comment by amihaiemil** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/jpeek/issues/100#issuecomment-362893860))

@amihaiemil This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#100](https://github.com/yegor256/jpeek/issues/100) is [in scope](http://datum.zerocracy.com/pages/policy.html#14) for 13days
  * The role is `DEV`
  * The job is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) for [2days](http://datum.zerocracy.com/pages/policy.html#8)
  * There is a monetary reward attached
  * @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) will get 50% of their hourly rate for this job
  * The job doesn't have any [impediments](http://datum.zerocracy.com/pages/policy.html#9)
  * The [budget](http://datum.zerocracy.com/pages/policy.html#4) is 30 minutes/points
  * These users are banned and won't be assigned:
    * @llorllale[/z](https://www.0crat.com/u/llorllale): This user reported the ticket

**Comment by amihaiemil** [comment]
@llorllale Does that comment still apply to the current XSL-based architedture?

``I think that we should introduce a new class which will do exactly that -- fetch the list of types from a method signature`` -- where should this class be? Where exactly are the method arguments counted now? I had a look in the project but it's not so clear :)

**Comment by amihaiemil** [comment]
@llorllale I had a look now and found the following: Skeleton doesn't actually count any arguments, it just lists them in XML. E.g. here is a method with 2 arguments, 1 Integer and 1 int:

```xml
<method abstract="false"
    ctor="false" desc="(Ljava/lang/Integer;I)V"
    name="test" public="true" static="false"
>
    <return>V</return>
    <args>
        <arg type="Ljava/lang/Integer">?</arg>
        <arg type="I">?</arg>
    </args>
</method>
```

So, question is: should it be the job of the Skeleton to count/differenciate the arguments or not? I would say no, since the skeleton's job should be representing the class *as is* in XML, with no altering.

Are we sure that all the metrics that are using the Skeleton are fine with ignoring the difference between ``int`` and ``Integer``? Shouldn't we leave it like this and let the metric decide what it needs?

**Comment by llorllale** [comment]
@amihaiemil 

> should it be the job of the Skeleton to count/differenciate the arguments or not? I would say no, since the skeleton's job should be representing the class as is in XML, with no altering.

Not coalescing them in `skeleton.xml` would lead to even more duplicate code in the XSL metrics. See the existing implementations - all have duplicate code for "should I include ctors or not?". We'll also start seeing "should I include nonpublic methods or not?" (see #101).

> Are we sure that all the metrics that are using the Skeleton are fine with ignoring the difference between int and Integer? Shouldn't we leave it like this and let the metric decide what it needs?

I'm not aware of any metric that delves into this level of detail. 

What I understand is that these ideas - no ctors, only public methods, coalesce primitives and their autoboxed types - are *opinions* introduced by us in jpeek, and are independent of the metrics.

**Comment by llorllale** [comment]
@amihaiemil btw, even though I didn't specify in the description, I think this should be configurable as well.

**Comment by llorllale** [comment]
@amihaiemil now you've got me thinking

Should we then also coalesce class attributes? This would have effects on how many `<op>` we list for methods that access attributes. 

So the scope of the problem is increasing with no clear benefits. In fact, I can already visualize some metrics changing value from some real number to `NaN` just from this change.

@yegor256 I'm starting to have doubts on this idea. Since you initially expressed the opinion, does it still hold for you?

**Comment by yegor256** [comment]
@llorllale @amihaiemil how about we create one more layer between `skeleton.xml` and `LCOM.xsl`, which will include a number of XSL stylesheets. One stylesheet will merge/coalesce types. Another one will remove ctors. Another one will remove static methods, and so on. That will help us get rid of code duplication inside XSL and will make the entire architecture composable: we can use certain XSL layers or ignore them. WDYT?

**Comment by amihaiemil** [comment]
@yegor256 Sounds good to me, I'll look into it :)

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by amihaiemil** [comment]
@0crat waiting

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/yegor256/jpeek/issues/100#issuecomment-364638567))

@amihaiemil The impediment for [#100](https://github.com/yegor256/jpeek/issues/100) was registered successfully by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by amihaiemil** [comment]
@llorllale can you close this now? :)

**Comment by llorllale** [comment]
hey @amihaiemil 
I see the XSL chain solution merged (which looks very nice by the way).

But is the problem reported in this ticket actually solved? Are primitives and their autoboxed counterparts being merged into one or the other type in some XSL?

**Comment by amihaiemil** [comment]
@llorllale no, with everything (initial inspection, discussion with you and the architect, implementing the first solution and then changing to this XlsChain at code review), the 30min were far exceeded. I left a todo for continuing implementation. Now we just need the stylesheets, the code architecture is in place :)

**Comment by amihaiemil** [comment]
@llorllale the ticket 116 is the next task here. 0pdd closed it but it is a mistake, i think, it should be open.

**Comment by amihaiemil** [comment]
@llorllale or maybe Yegor did some implementation for 116 and that is why it is closed (I see some commits referenced). Nevertheless, that is the next from here, I think this one can be closed :)

**Comment by llorllale** [comment]
@amihaiemil @yegor256 correct me if I'm wrong but the commit that closed #166 did not solve this ticket

**Comment by amihaiemil** [comment]
@llorllale this ticket was solved, from my point of view. We discussed and implemented code in such a way that the next steps (implementing stylesheets to extend the skeleton) are easily done in future tickets. And #166 is the ticket where implementation should continue -- whether it was closed by mistake or not should be discussed and solved there and this ticket should be closed.

**Comment by llorllale** [comment]
@amihaiemil alright, I opened #167 for this exact issue again :)

Thank you

**Comment by 0crat** [comment]
Order was successfully finished: +30 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+480](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
The job [#100](https://github.com/yegor256/jpeek/issues/100) is now out of scope

**Comment by yegor256** [comment]
@llorllale @amihaiemil I fixed the puzzle in #166 and removed it from the source code. Now we have XSL layers. We have just two of them: for removal of ctors and remove of static methods. If you need more, feel free to report bugs/tickets.
